### ISSUE TITLE
Deployment Fails to Complete - Hangs on Publishing to Azure...

### ISSUE THREAD

**Issue by mlipinski** [microsoft/botframework-solutions#2116]
#### What project is affected?

Virtual Assistant Template

#### What language is this in?

C#

#### What happens?

It hangs on publishing to Azure, it does seem to update the bot in Azure but neither locally nor in Azure can I get a response from the bot.

#### What are the steps to reproduce this issue?

Deploy the bot.

#### What were you expecting to happen?

Deployment and publish script complete successfully and able to access the virtual assistant bot in Emulator and in Azure Test in Webchat.

#### Can you share any logs, error output, etc.?

Yes, I can.
![PublishError](https://user-images.githubusercontent.com/5796298/63024834-00c77200-be76-11e9-80bd-2eb7bb4e5b2d.GIF)
[deploy_log.txt](https://github.com/microsoft/botframework-solutions/files/3501553/deploy_log.txt)
[publish_log.txt](https://github.com/microsoft/botframework-solutions/files/3501554/publish_log.txt)
[deploy_cognitive_models_log.txt](https://github.com/microsoft/botframework-solutions/files/3501555/deploy_cognitive_models_log.txt)

#### Any screenshots or additional context?

![PublishError](https://user-images.githubusercontent.com/5796298/63024912-26547b80-be76-11e9-9dd6-04b59acdcdb4.GIF)

**Comment by mlipinski** [comment]
Update, I can get to the bot via VS and the emulator after removing appid and apppw from the emulator and the appsettings.json file.

**Comment by darrenj** [comment]
Hi - the script uses the standard azure web app publishing mechanism, could you let us know the AZ version you are using please?

```
C:\Users\darrenj>az -v
azure-cli                         2.0.65 *
```

This will help us try and repro if there is a change in behaviour for az publish.

That said, creation of the azure resources has succeeded. One thing you can try to repro the issue would be to run publish directly, e.g. from powershell run the following and see if that works, if not can you share the publish_log.txt ?

```
deployment\scripts\publish.ps1 -name EALN_VATriageDev -resourceGroup EALN_VATriageDev

**Comment by mlipinski** [comment]
Darren,

I put the publish log and the deploy log in the github issue when I created it.


[cid:image001.png@01D5540F.7C21F320]

Matt…

**Comment by darrenj** [comment]
The logs don't show anything particularly wrong and I've just performed a full deployment to validate: 

![image](https://user-images.githubusercontent.com/33519520/63174654-564c7c00-c03a-11e9-983b-4e96818b7817.png)

I have just noticed this warning in your log above `command group bot is in preview it may be changed/removed in a future release` - This I'm 99.9% sure relates to you having a **very** old AZ version.

Can I confirm that you followed all the [pre-requsities](https://microsoft.github.io/botframework-solutions/tutorials/csharp/create-assistant/2_download_and_install/) specifically in this case the AZ version?

Also ensure you have the latest CLI tools as per the npm commands provided.

**Comment by mlipinski** [comment]
Darren,

I am confused on how this is a very old version.

https://docs.microsoft.com/en-us/cli/azure/install-azure-cli-windows?view=azure-cli-latest

This shows 2.0.71 is the most current version and I am on 2.0.70.

What am I missing here?

Yes, I did follow the pre-requisites before I deployed all this only a week ago.

Matt…

**Comment by dmvtech** [comment]
I've seen issues where PowerShell won't quite refresh the terminal when it's done (often happening with other, external commands). I've worked around this before by putting another command after it, such as:

`deployment\scripts\publish.ps1 -name EALN_VATriageDev -resourceGroup EALN_VATriageDev;Write-Host "The publishing should be complete now"`

But I don't quite think this is the case here as your bot doesn't actually work/respond.

How much time have you given the publishing step to complete?

**Comment by mlipinski** [comment]
I have left it for at least 45 minutes or more, I know it shouldn’t take that long.

**Comment by PurnaChandraPanda** [comment]
I have 2.0.71 az CLI. And, I see the same problem. It goes till "publishing to Azure", and then an hour passes but "done" message never appears for "deploy.ps1" - both virtual assistant/ skills deployment. However, we can forcefully exit command line session and do the normal debug work as it is. I am not sure where the problem is. Any idea why it could be happening?

**Comment by mlipinski** [comment]
I would like to deploy a few more bots but kind of on hold until I get this issue resolved.

**Comment by dmvtech** [comment]
Can you try running the deployment in _prompt mode_? i.e. instead of giving it each value for each parameter for one line, just run deploy.ps1 and then supply each value as it prompts for them. This is not a solution, but as a troubleshooting step. I'm curious to see if we can get more information by going that route.

**Comment by darrenj** [comment]
Could you also confirm the PowerShell Core version you have (pwsh -v). I've just double checked by upgrading to the latest AZ CLI to see if there was a regression and it's completed OK (within seconds).

Another potential workaround whilst we investigate what has regressed here is to right click and publish your solution to Azure via Visual Studio. Select the app-service that's been created and it should publish fine.

The publish script is a pretty simple wrapper around these two commands:

```
az bot prepare-deploy --lang Csharp --code-dir $projFolder --proj-file-path $projFile.name --output json | Out-Null
```

```
az webapp deployment source config-zip `
		--resource-group $resourceGroup `
		--name $name `
		--src $zipPath `
        --output json
```

If you have time you could try running the second command from the project directory manually - you should already have a code.zip file present and this will enable you to run against the az webapp command directly.

**Comment by mlipinski** [comment]
6.2.2 for Powershell

This only seems to happen when you include the App registration id as a parameter and not when its auto generated.  I don’t always have rights to auto generate, but yesterday I did.  This completed just fine with those rights.  When I have an app-id as a parameter it seems to hang.

--appid

Matt…

**Comment by darrenj** [comment]
Thanks - this helps. I've just gone through with latest pwsh and az CLI and not able to repro so rules out something with AZ CLI I suspect.

![image](https://user-images.githubusercontent.com/33519520/63348453-31fde180-c351-11e9-991d-313959532149.png)

I'm just re-testing with a manually created AppId. @PurnaChandraPanda can you confirm if you are also using the AppId and AppPassword parameter  when running the deployment script?

**Comment by darrenj** [comment]
Odd, using `.\Deployment\Scripts\deploy.ps1 -appId APP_ID -appPassword PASSWORD` worked for me. I'll install pwsh 6.2.2 (from my 6.2.1) to see if that makes any difference.

I'd also like to verify your publish.ps1 script is the same as [this](https://github.com/microsoft/botframework-solutions/blob/master/templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Deployment/Scripts/publish.ps1). It should be if you downloaded the template within the last month or so but wanted to double check.

If not, see [this](https://microsoft.github.io/botframework-solutions/reference/virtual-assistant/deploymentscripts/#how-do-i-update-my-local-deployment-scripts-with-the-latest)

**Comment by darrenj** [comment]
One other aspect that I want to highlight when your testing a Bot deployment. If you've had to create your own AppId it's imperative to ensure the tenant configuration is correct otherwise you will get AppId/Secret issues. This shouldn't affect deployment though!

See [these instructions](https://microsoft.github.io/botframework-solutions/reference/knownissues/#my-microsoft-app-registration-could-not-be-automatically-provisioned) for the configuration required for manually created Azure Applications - specifically **Accounts in any organizational directory**

**Comment by mlipinski** [comment]
I am not sure I chose “Accounts in any organizational directory”,

I typically select the first one in the list as I recall, but since I no longer have those rights I don’t recall which one that is.

Matt…

**Comment by mlipinski** [comment]
Yes, now I will say.  I copied and pasted that file to update it.  I am not 100% sure when I ran it the first time that it was the same version.

Matt…

**Comment by darrenj** [comment]
OK, to update - pwsh 6.2.2 worked fine for me on Windows 10 when specifying an appid and password to the deployment script. I'm not quite sure what else could be causing the publishing "hang" that you are experiencing. Also validated this on a Max/OSx.

So this is what I'd propose as next steps:

- [ ] Ensure "Accounts in any organizational directory" must be set on a manually created Azure AD application otherwise the Azure Bot Service will not be able to validate AppId/Secret. https://aka.ms/vamanualappcreation

- [ ] Validate the local project builds successfully. 

- [ ] Go to the Kudu site for your deployment - https://YOUR_BOT.scm.azurewebsites.net/DebugConsole. Click Site, Deployment and the most recent GUID to get the log file (share if possible)

- [ ] Publish manually via Visual Studio (selecting your Azure Web App in the Resource Group) and see if that works. If the Azure AD Application tenant setting discussed above is set, you should find Test in Web Chat (Azure Portal) and Emulator works as expected.

- [ ] Ensure you do have the [latest deployment scripts](https://microsoft.github.io/botframework-solutions/reference/virtual-assistant/deploymentscripts/#how-do-i-update-my-local-deployment-scripts-with-the-latest)

- [ ] Identify if executing `deployment\scripts\publish.ps1 -name EALN_VATriageDev -resourceGroup EALN_VATriageDev` directly repros the "hang issue" (this is called by the deploy script)

- [ ] If so, try adding  `--verbose` to the `az webapp deployment source config-zip` line in publish.ps1 to see if that yields anything further and share the log file if so. Alternatively you can execute the az command interactively within a powershell console.

**Comment by mickallen** [comment]
I am, also getting the exact same issue, I have followed all the pre-requisites did everything today so definitely latest versions. 

I will give the steps above a try.

**Comment by lauren-mills** [comment]
Closing due to inactivity. Please reopen as needed.

**Comment by hiramfernandes** [comment]
Not sure this is still relevant, but I went through the same issue and, in my case, what I found out was that there was a bunch of files created under {ProjectName}/Properties/PublishProfiles, most of them with the same name as the publish profile I was trying to create, which kind of indicates a failed attempt to create a publish profile and got stuck during the process.

What solved my issue was to manually delete all of those files and then the publish profile got created almost instantly. It looks like a file conflict issue that the Publish creation tool wasn't able to resolve.
